### PR TITLE
Add codegen test to C# distribtests.

### PR DESCRIPTION
**Pull request by jtattermusch**
This is to provide better tests for https://github.com/grpc/grpc/issues/14975.
- upgrade the distribtest dockerimages to the freshest mono installation (and remove some old hacks that are no longer necessary)
- remove the wheezy distribtest as debian wheezy support ended in May 2018
- add some basic codegen test in C# distribtest (currently only on linux and macos) - this also check that the executable flag is set for files extracted from a nuget package.

### CODE REVIEW

**Comment by jtattermusch** [comment]
Started artifact-package-distribtest run here: https://sponge.corp.google.com/target?id=d1435d57-41a2-46ef-93c4-f31ece1d6ba9&target=grpc/core/experimental/grpc_build_artifacts_multiplatform

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
The centos7 test is failing: https://sponge.corp.google.com/target?id=cef4d3a8-f3ae-4d83-9d5a-f8950b547105&target=github/grpc&searchFor=&show=ALL&sortBy=STATUS

File: test/distrib/csharp/test_codegen/test_codegen.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+# Copyright 2015 gRPC authors.
```
Comment (mehrdada) [inline test/distrib/csharp/test_codegen/test_codegen.sh]:
```
# Copyright 2018 The gRPC Authors
```

File: test/distrib/csharp/test_codegen/test_codegen.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  PLATFORM_ARCH=macosx_x64
+elif [ "$(getconf LONG_BIT)" == "32" ]
+then
+  PLATFORM_ARCH=linux_x86
+fi
+
+PROTOC="../packages/Grpc.Tools.__GRPC_NUGET_VERSION__/tools/${PLATFORM_ARCH}/protoc"
+PLUGIN="../packages/Grpc.Tools.__GRPC_NUGET_VERSION__/tools/${PLATFORM_ARCH}/grpc_csharp_plugin"
+
+$PROTOC --plugin="protoc-gen-grpc=${PLUGIN}" --csharp_out=. --grpc_out=. -I . testcodegen.proto
```
Comment (mehrdada) [inline test/distrib/csharp/test_codegen/test_codegen.sh]: Quotes in the above assignment aren't very useful. Instead, we should use `"$PROTOC"` here.

File: test/distrib/csharp/test_codegen/test_codegen.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+ls *.cs
```
Comment (mehrdada) [inline test/distrib/csharp/test_codegen/test_codegen.sh]: Shall we be more specific with the required filenames?

Comment (jtattermusch) [reply test/distrib/csharp/test_codegen/test_codegen.sh]: We are mostly relying on protoc returning exit code 0.

File: test/distrib/csharp/test_codegen/test_codegen.sh (hunk 2 above)
Comment (jtattermusch) [reply test/distrib/csharp/test_codegen/test_codegen.sh]: Done

File: test/distrib/csharp/test_codegen/test_codegen.sh (hunk 1 above)
Comment (jtattermusch) [reply test/distrib/csharp/test_codegen/test_codegen.sh]: Done

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
Passing package builds and distribtests are here (at 5def921):
https://sponge.corp.google.com/target?id=417030be-2096-4532-8dce-29f5ef6693f3&target=grpc%2Fcore%2Fexperimental%2Fgrpc_build_packages_multiplatform&searchFor=
https://sponge.corp.google.com/target?id=c488394a-136e-454c-9136-9f89f2a06a6f&target=grpc%2Fcore%2Fexperimental%2Fgrpc_distribtests_multiplatform&searchFor=

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
@matt-kwong  for owners approval.

**Pull request merged**